### PR TITLE
Fix test-install by moving tests into module

### PR DESCRIPTION
**Pull request by nelson-liu**
Goals of this PR:
1. [x] 1. Prevent tests from being installed as a module in `python setup.py` . Currently, pip copies them to `site-packages`, which is definitely not intended behavior.
2. [x] 2. installation method-invariant way to load and run tests, especially those that depend on files outside of the module (e.g. `scripts` or `tutorials`). The issue is that files can be in different places depending on if allennlp is installed with `pip` or `setup.py install`.

For 2: an easy solution might be to, when running the `test-install` command, simply download and cache a copy of the allennlp source for the appropriate version at `~/.allennlp` . Note that this wouldn't work for non-release versions, but I think that's a good thing --- if you're working on master or developing features, you shouldn't be using `test-install` anyway.

Thoughts? I'm not sure how i feel about the above solution, so feedback would be great before i write the code.

----------------------------------------------------------------------------------

*Edit*: another solution is to just move the tests into the `allennlp` module (this is a fairly standard way of structuring projects anyway). The downside to this is that it still doesn't enable using `allennlp test-install` to run tests in `tutorials`, as well as the tests that depend on files outside the module (e.g. SRL test that compares output of evaluator with a perl script). But maybe we shouldn't be running those tests anyway in `test-install`? or we can move the perl script into the tests and then just not run the notebook tests, since we'll be rewriting them anyway.

This solution is rather time-intensive, though, since it involves fixing all of the paths that are hardcoded into the tests. Should not be too hard to write a script to go through and change the paths, minimizing human labor to just a check up...

edit: fixes https://github.com/allenai/allennlp/issues/1228

### CODE REVIEW

**Comment by nelson-liu** [comment]
Another thing we could do is move the tests within the `allennlp` module...I know of at least a few others packages that do this (e.g., [scikit-learn](https://github.com/scikit-learn/scikit-learn/tree/master/sklearn/tests) and [pandas](https://github.com/pandas-dev/pandas/tree/master/pandas/tests). Maybe this solution is a bit drastic? I don't think too much depends on where the tests are actually located (maybe some of the CI scripts? But `py.test` in the root should continue to work)

**Comment by nelson-liu** [comment]
I went ahead and made the change, was a lot less work than I anticipated. even if we don't want to move tests within the module, a lot of these commits can be cherry-picked for an eventual window-compatibility PR.

maybe best to look at this after EMNLP...

**Comment by schmmd** [comment]
@nelson-liu what's the difference between `tools` and `scripts`?

**Comment by nelson-liu** [comment]
1) `scripts` are standalone, one-off entities, while `tools` are core to some other component of the package (e.g. `evalb` is essential to the constituency parser metric, and the srl perl eval script is used in the tests). In general, if some outside tool/script is necessary to make a component of the library function properly, it should go into `tools`.

2) since scripts are standalone, they go outside of the package. Tools are explicitly called by things inside the package, so they should go inside the package as well. I named them differently to make it easier to distinguish.

**Comment by nelson-liu** [comment]
(let me know if there's anything I can do to make this easier to review, as well).

**Comment by schmmd** [comment]
@nelson-liu I think the names are fine.  A brief (1 sentence) README.md in each folder would be great, defining the role of each.  Overall, this seems like a lot of improvements, but also a lot of changes.  As it touches the whole codebase I'm not sure if I'm the best person to code review it.

File: allennlp/common/file_utils.py [hunk 1]
```diff
@@ -60,7 +61,7 @@ def filename_to_url(filename: str, cache_dir: str = None) -> Tuple[str, str]:
 
     return url, etag
 
-def cached_path(url_or_filename: str, cache_dir: str = None) -> str:
```
Comment (DeNeutoy) [inline allennlp/common/file_utils.py]: @joelgrus does this `Union`, and all the `str(path)` calls in this PR change how you feel about `Path`?

**Comment by DeNeutoy** [comment]
I'm still unsure whether it's desirable to have users run tests. Can you point me to some other python projects which do this?

**Comment by nelson-liu** [comment]
NumPy, SciPy let users run their tests through the CLI (https://stackoverflow.com/a/9200923/2544124):

```
import numpy
numpy.test()
```

(and accordingly for scipy)

Also [theano](http://deeplearning.net/software/theano_versions/0.8.X/install.html#testing-your-installation)

**Comment by schmmd** [comment]
For another motivating use case, I'd like to create a Docker image that downstream users can use to run Beaker.  It would look [like @cristipp 's](https://github.com/allenai/euclid/blob/master/src/main/python3/pipette/Dockerfile) and would install AllenNLP via `pip`.  Having the ability to test the pip distribution would be invaluable for making sure the image works as expected.

File: allennlp/tests/commands/test_install_test.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# pylint: disable=invalid-name,no-self-use
+import os
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.commands.test_install import _get_module_root
+
+
+class TestTestInstall(AllenNlpTestCase):
```
Comment (schmmd) [inline allennlp/tests/commands/test_install_test.py]: What is the motivation for this test?  A brief comment would be helpful.  I'm confused because it's a test, in the tests folder, that's checking to see whether the tests folder exists.

File: allennlp/tests/training/metrics/wikitables_accuracy_test.py [hunk 3]
```diff
@@ -29,7 +31,8 @@ def test_accuracy_is_scored_correctly(self):
         logical_form = ('((reverse fb:row.row.year) (fb:row.row.index (max '
                         '((reverse fb:row.row.index) (fb:row.row.league fb:cell.usl_a_league)))))')
 
-        wikitables_accuracy = WikiTablesAccuracy(table_directory='tests/fixtures/data/wikitables/')
+        print(os.getcwd())
```
Comment (schmmd) [inline allennlp/tests/training/metrics/wikitables_accuracy_test.py]: @nelson-liu is this an intentional `print`?

File: mypy.ini [hunk 4]
```diff
@@ -0,0 +1,4 @@
+[mypy]
+
+[mypy-allennlp.tests*]
+ignore_errors = True
```
Comment (schmmd) [inline mypy.ini]: Why is this necessary? Did `mypy` previously not run on our tests?

File: allennlp/tests/commands/test_install_test.py (hunk 2 above)
Comment (schmmd) [reply allennlp/tests/commands/test_install_test.py]: It looks like this is a rename rather than a new file.  While that doesn't change my confusion, you don't need to address it as part of this PR.

File: mypy.ini (hunk 4 above)
Comment (nelson-liu) [reply mypy.ini]: Yes, mypy was previously not being run on the testing code --- I thought this was intentional?

File: allennlp/tests/training/metrics/wikitables_accuracy_test.py (hunk 3 above)
Comment (nelson-liu) [reply allennlp/tests/training/metrics/wikitables_accuracy_test.py]: it's not, thanks for catching that!

File: mypy.ini (hunk 4 above)
Comment (schmmd) [reply mypy.ini]: Seems like you made the right change in this PR then.  I'm not sure what we want long term.

**Comment by schmmd** [comment]
@nelson-liu sounds great.  If you can move `srl-eval.pl` instead of creating a new copy, then I'll approve and merge.

File: allennlp/tests/commands/test_install_test.py (hunk 2 above)
Comment (nelson-liu) [reply allennlp/tests/commands/test_install_test.py]: I agree that this is quite confusing. The context is that when a user runs `test-install`, we have no idea where they're running it from, so we do an `os.chdir` to the module root in order to get all the paths to resolve properly.

The logic within `test-install` is pretty hard to test in its entirety ([see discussion here](https://github.com/allenai/allennlp/pull/1213#issuecomment-388897160)), so this test is basically a test for that `os.chdir` code, checking that we correctly find the path to `chdir` to.

does that help? I'll add this as a comment too.

**Comment by nelson-liu** [comment]
before you merge, i'd like to run some last sanity-check tests on the code in this branch (building it from github source with pip on a clean environment and seeing if everything works as expected). I'll let you know when that's done.

Maybe this is something you might want to add to CI as well? Not sure.

**Comment by DeNeutoy** [comment]
I was semi-serious about not liking the `Path` thing - I think it will just cause us to have loads of `str(path)` things everywhere and doesn't really add much. I'd kinda prefer it if this PR switched back to using string paths.... @joelgrus lets argue

**Comment by DeNeutoy** [comment]
Meh maybe i'm just being grumpy and un-modern - I didn't realise it might fix some of the windows compatability issues. If that's the case let's go for it.

File: allennlp/tests/commands/test_install_test.py (hunk 2 above)
Comment (schmmd) [reply allennlp/tests/commands/test_install_test.py]: Yes that helps!

File: mypy.ini (hunk 4 above)
Comment (matt-gardner) [reply mypy.ini]: @nelson-liu, do the mypy errors on the tests look helpful?  I'd guess that we don't care about them, because it's not as important to have clear type annotations in the tests, but if they look useful we could reconsider.

**Comment by matt-gardner** [comment]
@DeNeutoy, I agree that the only benefit I see to using the `Path` stuff is better windows compatibility.  Otherwise just using a string path is simpler and easier to read.

File: mypy.ini (hunk 4 above)
Comment (nelson-liu) [reply mypy.ini]: @matt-gardner here are the mypy errors: https://gist.github.com/nelson-liu/f94a96265973a390066f45399718fe5e . I guess they're not entirely useless? Perhaps we should address in another PR though, this one is already huge.

**Comment by nelson-liu** [comment]
I'll merge this when CI passes, unless there's anything further to be discussed?

I pip installed allennlp from this branch on an osx machine and a linux gpu machine, and `allennlp test-install` passed on both hosts.

**Comment by joelgrus** [comment]
to @DeNeutoy 's point, calling `str()` on `pathlib.Path` objects is a bad code smell. if the places we need those paths can't accept `pathlib.Path` object then I would rather just use strings.

(my ideal solution would be that those places accept Path objects tho)

File: mypy.ini (hunk 4 above)
Comment (matt-gardner) [reply mypy.ini]: Hmm, yeah, ok, I'll open that as an issue.  Thanks for doing all of this!

**Comment by nelson-liu** [comment]
thanks all for taking a look
